rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,825 @@
+
+**Proposal:** This RFC proposes to immediately remove these 6 operations from
+their current dialects and move them into a new dialect called `chlo_legacy`.
+
+StableHLO [doesn't guarantee](../docs/compatibility.md#out-of-scope)
+compatibility for these operations (see the "Unspecced features" clause).
+However, this RFC proposes to provide compatibility guarantees on exceptional
+basis given that they don't appear to involve a lot of work. More specifically,
+the proposal is to keep the `chlo_legacy` dialect in the StableHLO repository
+for 6 months from the date of its creation and only then remove it.
+
+<!-- markdownlint-disable line-length -->
+## <a id="p2" /> (P2) Ratify the existing convention for shape mismatches constituting undefined behavior
+<!-- markdownlint-enable line-length -->
+
+**Baseline:** As mentioned in [F7](#f7), there is broad consensus that shape
+mismatches in StableHLO operations should constitute undefined behavior, i.e.
+that guarding against shape mismatches should be made a producer responsibility.
+This has been a subject of many informal conversations as well as one of the
+discussion items
+[at the StableHLO dynamism breakout session](https://drive.google.com/drive/u/1/folders/1fGqq8Tcebhcwq1KJqAZPDgXksRdqRvM2)
+during the OpenXLA Dev Summit in April 2023.
+
+**Proposal:** This RFC proposes to ratify this convention as a StableHLO design
+principle, so that the folklore consensus becomes codified in the StableHLO
+specification.
+
+**Discussion:** A) The rationale for this proposal is that lower-level
+abstractions (e.g. HLO or Linalg) typically don't want to concern themselves
+with shape checks - they are focused on generating high-performance code, so
+they want shape checks to be handled at a higher level, i.e. at StableHLO
+or above.
+
+Furthermore, different producers have different requirements and different
+preferences for expressing shape checks (e.g. JAX's type system enables it to
+need fewer checks than TensorFlow's type system), so a specific way of
+performing shape checks don't look like something that should be standardized
+at the StableHLO level either.
+
+B\) From the implementation standpoint, this proposal would involve updating
+[the "Errors" section](../docs/spec.md#errors) of the specification, and
+changing the StableHLO dialect to implement the `ConditionallySpeculatable`
+interface, prohibiting speculation for StableHLO ops that involve dynamism
+([documentation](https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/)).
+
+<!-- markdownlint-disable line-length -->
+## <a id="p3" /> (P3) Ratify the existing convention for relaxed constraints already implemented in the StableHLO dialect
+<!-- markdownlint-enable line-length -->
+
+**Baseline:** At the moment, the StableHLO dialect supports
+[relaxed constraints](https://github.com/openxla/stablehlo/blob/8993ad54839add6648b88801f1d223b7f9bc2e58/stablehlo/dialect/Base.cpp#L102-L120)
```

Comment:
Is this true even for unranked if unranked is probably going out of scope?

Reply:
Yes, unranked types are also part of these relaxed constraints in the current StableHLO implementation. I didn't call them out in the example below though to avoid overloading it, but they are in scope of the proposed design principle "If a constraint for a specific operation cannot be evaluated at compile time because it involves an unknown rank or an unknown dimension size, it gets deferred until the run time of the operation".